src/decode/ykpt/mod.rs

```diff
@@ -0,0 +1,61 @@
+//! The Yk PT trace decoder.
+
+use crate::{decode::TraceDecoder, errors::HWTracerError, Block, Trace};
+
+mod packet_parser;
+use packet_parser::PacketParser;
+
+pub(crate) struct YkPTTraceDecoder {}
+
+impl TraceDecoder for YkPTTraceDecoder {
+    fn new() -> Self {
+        Self {}
+    }
+
+    fn iter_blocks<'t>(
+        &'t self,
+        trace: &'t dyn Trace,
+    ) -> Box<dyn Iterator<Item = Result<Block, HWTracerError>> + '_> {
+        let itr = YkPTBlockIterator {
+            errored: false,
+            parser: PacketParser::new(trace.bytes()),
+        };
+        Box::new(itr)
+    }
+}
+
+/// Iterate over the blocks of an Intel PT trace using the fast Yk PT decoder.
+struct YkPTBlockIterator<'t> {
+    /// Error state. Set to true when an error occurs, thus invalidating the iterator.
+    errored: bool,
+    /// PT packet iterator.
+    parser: PacketParser<'t>,
+}
+
+impl<'t> Iterator for YkPTBlockIterator<'t> {
+    type Item = Result<Block, HWTracerError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // There was an error in a previous iteration.
+        if self.errored {
```

Comment:
I tend to prefer:

```rust
if !self.errored {
  // common case
} else {
  // uncommon case
}
```

on the basis that compilers *tend* to guess that the "true" branch is the common case.

Reply:
The idiom was copied from the [libipt decoder](https://github.com/ykjit/hwtracer/blob/9e517be439aaad9b5b8ffeac2b1b8d5f0860c6f2/src/decode/libipt/mod.rs#L119), where it makes more sense as the method is long and you don't want so much indented code.

I expect this function will grow too, but I wouldn't worry too much right now as we don't actually cover this code path yet. It's only implemented to satisfy a trait at the moment.

Reply:
The problem with perpetuating the bad idiom is that it gives us permission to perpetuate it again (and again (and again)). It would be better to fix this now IMHO.

Reply:
Fixed in [3567abf](https://github.com/ykjit/hwtracer/pull/105/commits/3567abfb722b6106c105a07bc9a0b4ceda47d3a4)